File: crates/pk-core/src/maze/lamp_puzzle.rs
//! Light puzzle authoring and runtime simulation — brazier and sealed loot vault placement.
//!
//! Port of `legacy/src/game/pinball-knight/lamp-puzzle.ts` (174 lines).
//!
//! PORTS: `lamp-puzzle.ts`

use std::sync::RwLock;

use crate::flow_field::bfs_distances;
use crate::grid::{at, idx, tile_center, Grid, T_FLOOR};
use crate::rng::Mulberry32;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LampPuzzlePlan {
    /// Brazier tile positions (i, j)
    pub lamps: Vec<(i32, i32)>,
    /// The sealed chest vault tile position (i, j)
    pub vault: (i32, i32),
    /// Potion rewards unlocked upon puzzle completion
    pub loot: Vec<String>,
}

pub const LOOT_TABLES: [&[&str]; 5] = [
    &["gold", "gold", "health"],
    &["gold", "shield", "health"],
    &["gold", "haste", "gold"],
    &["rage", "gold", "health"],
    &["gold", "ballform", "health"],
];

pub const CHEST_UNLIT: u32 = 0x8a2020;
pub const CHEST_LIT: u32 = 0xd98020;
pub const CHEST_OPEN: u32 = 0xffe066;

#[derive(Debug, Clone, Default)]
pub struct LampPuzzleState {
    pub total: usize,
    pub lit: usize,
    pub unlocked: bool,
    pub vault_i: i32,
    pub vault_j: i32,
    pub vault_x: f64,
    pub vault_z: f64,
    pub loot: Vec<String>,
    pub open_t: f64,
}

static PUZZLE_STATE: RwLock<Option<LampPuzzleState>> = RwLock::new(None);

pub fn install_lamp_puzzle(plan: &LampPuzzlePlan, g: &Grid) {
    let c = tile_center(g, plan.vault.0, plan.vault.1);
    let state = LampPuzzleState {
        total: plan.lamps.len(),
        lit: 0,
        unlocked: false,
        vault_i: plan.vault.0,
        vault_j: plan.vault.1,
        vault_x: c.0,
        vault_z: c.1,
        loot: plan.loot.clone(),
        open_t: -1.0,
    };
    *PUZZLE_STATE.write().unwrap() = Some(state);
}

pub fn light_lamp() -> bool {
    let mut w = PUZZLE_STATE.write().unwrap();
    if let Some(ref mut pz) = *w {
        if !pz.unlocked {
            pz.lit += 1;
            if pz.lit >= pz.total {
                pz.unlocked = true;
                pz.open_t = 0.0;
                return true;
            }
        }
    }
    false
}

pub fn update_lamp_puzzle(dt: f64) {
    let mut w = PUZZLE_STATE.write().unwrap();
    if let Some(ref mut pz) = *w {
        if pz.unlocked {
            pz.open_t += dt;
        }
    }
}

pub fn dispose_lamp_puzzle() {
    *PUZZLE_STATE.write().unwrap() = None;
}

/// How many braziers a floor of this depth gets (3-5).
pub fn lamp_count_for(level: u32) -> usize {
    (3 + (level / 3) as usize).clamp(3, 5)
}

/// Authors a lamp puzzle for a floor, or returns None if the floor is too small or crowded.
pub fn author_lamp_puzzle(
    g: &Grid,
    start: (i32, i32),
    occupied: impl Fn(i32, i32) -> bool,
    rng: &mut Mulberry32,
    lamp_count: usize,
) -> Option<LampPuzzlePlan> {
    let d = bfs_distances(g, start.0, start.1);
    let mut max_d = 0;
    for &dist in &d {
        if dist >= 0 && dist > max_d {
            max_d = dist;
        }
    }
    if max_d < 8 {
        return None;
    }

    // Reachable, unoccupied floor with breathing room from start
    let mut cand = Vec::new();
    for j in 1..(g.h - 1) {
        for i in 1..(g.w - 1) {
            if at(g, i, j) != T_FLOOR {
                continue;
            }
            let dd = d[idx(g, i, j)];
            if dd < 4 {
                continue;
            }
            if occupied(i, j) {
                continue;
            }
            cand.push((i, j, dd as u32));
        }
    }

    if cand.len() < lamp_count + 3 {
        return None;
    }

    // Vault: a far, open tile (all 4 cardinals floor)
    let is_open_tile = |i: i32, j: i32| -> bool {
        at(g, i + 1, j) == T_FLOOR
            && at(g, i - 1, j) == T_FLOOR
            && at(g, i, j + 1) == T_FLOOR
            && at(g, i, j - 1) == T_FLOOR
    };

    let mut far: Vec<(i32, i32, u32)> = cand
        .iter()
        .copied()
        .filter(|c| c.2 >= ((max_d as u32) / 2))
        .collect();
    far.sort_by(|a, b| b.2.cmp(&a.2));

    let vault_pos = far
        .iter()
        .find(|c| is_open_tile(c.0, c.1))
        .map(|c| (c.0, c.1))
        .or_else(|| far.first().map(|c| (c.0, c.1)))
        .unwrap_or((cand.last().unwrap().0, cand.last().unwrap().1));

    // Braziers: spread across reachable floor with minimum separation
    let mut lamps = Vec::new();
    let mut available: Vec<(i32, i32, u32)> = cand
        .into_iter()
        .filter(|c| (c.0, c.1) != vault_pos)
        .collect();

    let min_sep_sq = 9; // ~3 tile separation
    while lamps.len() < lamp_count && !available.is_empty() {
        let pick_idx = (rng.next_f64() * available.len() as f64) as usize % available.len();
        let candidate = available.swap_remove(pick_idx);

        let far_enough = lamps.iter().all(|l: &(i32, i32)| {
            let dx = candidate.0 - l.0;
            let dz = candidate.1 - l.1;
            dx * dx + dz * dz >= min_sep_sq
        });

        if far_enough {
            lamps.push((candidate.0, candidate.1));
        }
    }

    if lamps.len() < lamp_count {
        return None;
    }

    let loot_idx = (rng.next_f64() * LOOT_TABLES.len() as f64) as usize % LOOT_TABLES.len();
    let loot = LOOT_TABLES[loot_idx]
        .iter()
        .map(|s| s.to_string())
        .collect();

    Some(LampPuzzlePlan {
        lamps,
        vault: vault_pos,
        loot,
    })
}

File: ThreeJS/src/game/pinball-knight/maze/lamp-puzzle.ts
/**
 * LIGHT PUZZLE authoring (pure) — where the braziers and the sealed loot vault
 * go on a floor.
 *
 * The floor gets a **sealed vault** (a glowing chest in an open spot, far from
 * the start) and N unlit **braziers** scattered across the reachable floor.
 * Rolling the pinball knight over each brazier lights it; the last one opens the
 * vault (see dungeon/lamp-puzzle.ts for the runtime + reward). This module only
 * picks TILES — it never touches THREE or state — so it is fully unit-testable.
 *
 * Invariant-safe: it only READS the grid and only ever places things on
 * existing reachable FLOOR tiles (the vault is a chest, not a carved sealed
 * room — the "every floor tile reachable" pipeline invariant is untouched).
 */
import { type Grid, type TilePos, T_FLOOR, at, idx } from "./generator";
import type { PinballPartSpot } from "./decorate";
import { bfsDistances, bfsDistancesOwned } from "../engine/flow-field";

export interface LampPuzzlePlan {
  /** Brazier spots (already PinballPartSpots so they inject into plan.parts). */
  lamps: PinballPartSpot[];
  /** The sealed chest tile. */
  vault: TilePos;
  /** Potion ids paid out when the vault opens (see items.ts POTIONS). */
  loot: string[];
}

/**
 * Loot tables — all POTION ids (ground items spawn as kind:"potion").
 *
 * Exported so `potion-supply.test.ts` can count this as a supply route: a vault
 * payout is a real way to obtain a potion, and a guard that only knew about the
 * floor pool and the shops would call these ids unreachable.
 */
export const LOOT_TABLES: readonly (readonly string[])[] = [
  ["gold", "gold", "health"],
  ["gold", "shield", "health"],
  ["gold", "haste", "gold"],
  ["rage", "gold", "health"],
  ["gold", "ballform", "health"],
];

/** How many braziers a floor of this size/depth gets (3-5). */
export function lampCountFor(level: number): number {
  return Math.max(3, Math.min(5, 3 + Math.floor(level / 3)));
}

/**
 * Pick the puzzle's tiles, or null if the floor is too small / too crowded to
 * host one cleanly. `occupied(i,j)` reports tiles already carrying placed
 * content (parts, spawns, items, props, torches, stairs, start) — nothing
 * lands on top of them. When `bossSpot` is provided, the vault is sited
 * immediately adjacent to the boss in the boss chamber.
 */
export function authorLampPuzzle(
  g: Grid,
  start: TilePos,
  occupied: (i: number, j: number) => boolean,
  rng: () => number,
  lampCount: number,
  bossSpot?: TilePos | null,
): LampPuzzlePlan | null {
  const d = bfsDistancesOwned(g, start.i, start.j); // held while placing
  let maxD = 0;
  for (let k = 0; k < d.length; k++) if (d[k] > maxD) maxD = d[k];
  if (maxD < 8) return null; // too small to bother

  // Reachable, unoccupied floor with a little breathing room from the start.
  const cand: Array<{ i: number; j: number; d: number }> = [];
  for (let j = 1; j < g.h - 1; j++) {
    for (let i = 1; i < g.w - 1; i++) {
      if (at(g, i, j) !== T_FLOOR) continue;
      const dd = d[idx(g, i, j)];
      if (dd < 4) continue;
      if (occupied(i, j)) continue;
      cand.push({ i, j, d: dd });
    }
  }
  if (cand.length === 0) return null;

  // Vault: an open tile (all four cardinals floor → reads as a chamber, not
  // a wall nook).
  const openTile = (t: { i: number; j: number }): boolean =>
    at(g, t.i + 1, t.j) === T_FLOOR &&
    at(g, t.i - 1, t.j) === T_FLOOR &&
    at(g, t.i, t.j + 1) === T_FLOOR &&
    at(g, t.i, t.j - 1) === T_FLOOR;

  let vault: TilePos;
  if (bossSpot) {
    // When bossSpot is specified, site the vault right in the boss chamber.
    const bossCand = cand.slice().sort((a, b) => {
      const da = Math.abs(a.i - bossSpot.i) + Math.abs(a.j - bossSpot.j);
      const db = Math.abs(b.i - bossSpot.i) + Math.abs(b.j - bossSpot.j);
      return da - db;
    });
    // Prefer an open tile within 1..3 tiles of the boss, or the closest candidate.
    const closeOpen = bossCand.find((c) => {
      const dist = Math.abs(c.i - bossSpot.i) + Math.abs(c.j - bossSpot.j);
      return dist >= 1 && dist <= 3 && openTile(c);
    });
    const best = closeOpen ?? bossCand[0];
    vault = { i: best.i, j: best.j };
  } else {
    // Fallback: a far, open tile.
    const far = cand.filter((c) => c.d >= maxD * 0.5).sort((a, b) => b.d - a.d);
    const vaultC = far.find(openTile) ?? far[0] ?? cand[cand.length - 1];
    vault = { i: vaultC.i, j: vaultC.j };
  }

  // Braziers: spread across the reachable floor (separation scales with size),
  // never hugging the vault or each other.
  const sep = Math.max(6, Math.floor(maxD / 6));
  const lampsT: TilePos[] = [];
  for (const c of shuffle(cand, rng)) {
    if (lampsT.length >= lampCount) break;
    if (Math.abs(c.i - vault.i) + Math.abs(c.j - vault.j) < 3) continue;
    if (bossSpot && Math.abs(c.i - bossSpot.i) + Math.abs(c.j - bossSpot.j) < 3) continue;
    if (lampsT.some((l) => Math.abs(l.i - c.i) + Math.abs(l.j - c.j) < sep)) continue;
    lampsT.push({ i: c.i, j: c.j });
  }

  const lamps = lampsT.map(
    (t): PinballPartSpot => ({ i: t.i, j: t.j, kind: "lamp", dirI: 0, dirJ: 0, dir2I: 0, dir2J: 0 }),
  );
  const loot = [...LOOT_TABLES[Math.floor(rng() * LOOT_TABLES.length)]];
  return { lamps, vault, loot };
}

function shuffle<T>(a: readonly T[], rng: () => number): T[] {
  const arr = a.slice();
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}
